src/tools/miri/src/shims/unix/linux/sync.rs

```diff
@@ -152,12 +158,16 @@ pub fn futex<'tcx>(
             //
             // Thankfully, preemptions cannot happen inside a Miri shim, so we do not need to
             // do anything special to guarantee fence-load-comparison atomicity.
-            this.atomic_fence(AtomicFenceOrd::SeqCst)?;
-            // Read an `i32` through the pointer, regardless of any wrapper types.
-            // It's not uncommon for `addr` to be passed as another type than `*mut i32`, such as `*const AtomicI32`.
-            let futex_val = this.read_scalar_atomic(&addr, AtomicReadOrd::Relaxed)?.to_i32()?;
-            if val == futex_val {
+            let futex_val = this.read_scalar_atomic(&addr, AtomicReadOrd::SeqCst)?.to_u32()?;
```

Comment:
> SeqCst writes and SeqCst fences are very much not equivalent.

I have no idea. Could you please explain it or provide some materials?

Reply:
I'm afraid a full introduction to the C++ memory model is beyond the scope of this thread. Mara wrote a book about it, available at https://marabos.nl/atomics/, but I don't know if it goes into the fact that SeqCst fences + relaxed accesses are not equivalent to SeqCst accesses -- that is *really advanced*, and I don't know any place that thoroughly explains it.

Please keep the fence, and make all the reads/writes `Relaxed`, like it was before. Carefully read the comment to ensure the fence is put in the right place, given that there are some new accesses being added here.

Reply:
What about `write_scalar_atomic` to `&addr`? Do I need to put a fence after it?

Reply:
I don't know. The old logic was carefully figured out by a bunch of people together. Any adjustment of it will require someone to really dig into this, understand the logic, and make sure the adjustment makes sense.

Reply:
🤯

Reply:
Yeah sorry it's complicated. :/  I can try to help, but my time is very limited.

If there is some way to land this without doing the Miri change, that would also work. E.g. we could keep using `futex::Mutex` instead of `pi_futex::Mutex` when `cfg(miri)` is set.

Reply:
I think it's correct now in my latest commit. Someone can report that if it's not, but for now I assume it is as it passed tests.

Reply:
Tests passing doesn't mean it is correct, it just means it is good enough to not blow up. ;)

If you want to land the Miri changes, we'll definitely need a test as requested [here](https://github.com/rust-lang/rust/pull/131584#issuecomment-2408711620), in particular a version of `concurrent_wait_wake` for PI futexes.

Reply:
> If you want to land the Miri changes, we'll definitely need a test as requested https://github.com/rust-lang/rust/pull/131584#issuecomment-2408711620, in particular a version of concurrent_wait_wake for PI futexes.

I've added one in the latest commit. Plz have a look :smiley: